ably.ts.d

```diff
@@ -0,0 +1,394 @@
+
+namespace PresenceAction {
+  export type ABSENT = "absent";
+  export type PRESENT = "present";
+  export type ENTER = "enter";
+  export type LEAVE = "leave";
+  export type UPDATE = "update";
+}
+type PresenceAction = PresenceAction.ABSENT | PresenceAction.PRESENT | PresenceAction.ENTER | PresenceAction.LEAVE | PresenceAction.UPDATE;
+
+namespace StatsIntervalGranularity {
+  export type MINUTE = "minute";
+  export type HOUR = "hour";
+  export type DAY = "day";
+  export type MONTH = "month";
+}
+type StatsIntervalGranularity = StatsIntervalGranularity.MINUTE | StatsIntervalGranularity.HOUR | StatsIntervalGranularity.DAY | StatsIntervalGranularity.MONTH;
+
+// Interfaces
+
+interface ClientOptions extends AuthOptions {
+  autoConnect?: boolean;
+  clientId?: String;
+  defaultTokenParams?: TokenParams;
+  echoMessages?: boolean;
+  environment?: String;
+  logHandler?: any;
+  logLevel?: any;
+  port?: number;
+  queueMessages?: boolean;
+  restHost?: String;
+  realtimeHost?: String;
+  fallbackHosts?: Array<String>
+  recover?: String
+  tls?: boolean;
+  tlsPort?: number;
+  useBinaryProtocol?: boolean;
+}
+
+interface AuthOptions {
+  authCallback?: (data: String | TokenDetails | TokenRequest) => void;
+  authHeaders?: Array<any>;
+  authMethod?: String;
+  authParams?: Array<any>;
+  authUrl?: String;
+  key?: String;
+  queryTime?: boolean;
+  token?: String | TokenDetails;
+  tokenDetails?: TokenDetails;
+  useTokenAuth?: boolean;
+}
+
+interface TokenParams {
+  capability: String;
+  clientId?: String;
+  nonce?: String;
+  timestamp: Number;   // Time?
+  ttl: Number;
+}
+
+interface CipherParams {
+  algorithm: String;
+  key: any;
+  keyLength: number;
+  mode: String;
+}
+
+interface ErrorInfo {
+  code: number;
+  message: String;
+  statusCode: number;
+}
+
+interface StatsMessageTypes {
+  all: number,
+  messages: number,
+  presence: number;
+}
+
+interface StatsRequestCount {
+  failed: number,
+  refused: number,
+  succeeded: number
+}
+
+interface StatsResourceCount {
+  mean: number,
+  min: number,
+  opened: number,
+  peak: number,
+  refused: number
+}
+
+interface StatsConnectionTypes {
+  all: number,
+  plain: number,
+  tls: number
+}
+
+interface StatsMessageTraffic {
+  all: number,
+  realtime: number,
+  rest: number,
+  webhook: number
+}
+
+interface TokenDetails {
+  capability: String;
+  clientId: String;
+  expires: number; // Time
+  issued: number; // Time
+  token: String;
+}
+
+interface TokenRequest {
+  capability: String;
+  clientId: String;
+  keyName: String;
+  mac: String;
+  nonce: String;
+  timestamp: number; // Time
+  ttl: number;
+}
+
+interface ChannelOptions {
+  cipher: any;
+}
+
+interface RestPresenceHistoryParams {
+  start: number;
+  end: number;
+  direction: String;
+  limit: number;
+}
```

Comment:
similarly, users don't need to provide all of these, they can give any subset. If these aren't marked optional, won't that mean that their program won't compile unless they give all? Or am I misunderstanding how ts works?